### PR TITLE
Move smell-related classes out of /core to /smells.

### PR DESCRIPTION
**Pull request by troessner**
Fixes the last part of #517 and move SmellConfiguration and SmellRepository to lib/reek/smells.

Both classes are related to smells, but not to anything else in /core AFAIS.

### CODE REVIEW

**Comment by chastell** [comment]
I’m ok with this one, but I remember @mvz having concerns about moving `SmellRepository` under `Smells`, so I’m deferring this merge to him.

**Comment by mvz** [comment]
I was worried that the moving of SmellConfiguration might unnecessarily break our external API, undefined as it is, but it seems I was mistaken. I suppose it's mostly Examiner that is used externally.

File: lib/reek/smells/attribute.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 require_relative 'smell_detector'
-require_relative '../core/smell_configuration'
+require_relative '../smells/smell_configuration'
```
Comment (mvz) [inline lib/reek/smells/attribute.rb]: You can now remove the `../smells/` bit from the require here and in other files under `smells/`.

File: lib/reek/smells/attribute.rb [hunk 2]
```diff
@@ -31,7 +31,7 @@ def self.contexts # :nodoc:
       end
 
       def self.default_config
-        super.merge(Core::SmellConfiguration::ENABLED_KEY => false)
+        super.merge(Smells::SmellConfiguration::ENABLED_KEY => false)
```
Comment (mvz) [inline lib/reek/smells/attribute.rb]: Similarly, the `Smells` could be removed here.

File: lib/reek/smells/smell_repository.rb [hunk 3]
```diff
@@ -1,9 +1,9 @@
 require_relative '../smells'
```
Comment (mvz) [inline lib/reek/smells/smell_repository.rb]: Perhaps a next step would be to move the contents of `smells.rb` into here.

**Comment by mvz** [comment]
Good to go as far as I'm concerned after the requires and namespacing have been simplified.

File: lib/reek/smells/attribute.rb (hunk 1 above)
Comment (troessner) [inline lib/reek/smells/attribute.rb]: Turns out this require wasn't even needed.:)

File: lib/reek/smells/long_parameter_list.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 require_relative 'smell_detector'
-require_relative '../core/smell_configuration'
+require_relative '../smells/smell_configuration'
```
Comment (troessner) [inline lib/reek/smells/long_parameter_list.rb]: Same here, was not needed at all

File: lib/reek/smells/smell_repository.rb (hunk 3 above)
Comment (troessner) [inline lib/reek/smells/smell_repository.rb]: Good idea! Then you'd just have to require SmellRepository...

Comment (troessner) [inline lib/reek/smells/smell_repository.rb]: Let's make that a separate PR to keep things simple.

**Comment by troessner** [comment]
@mvz addressed all your comments. Merge?.:)

**Comment by troessner** [comment]
Seems like a random travis failure, restarting...

File: lib/reek/smells/smell_repository.rb (hunk 3 above)
Comment (mvz) [inline lib/reek/smells/smell_repository.rb]: Agreed.

**Comment by mvz** [comment]
Looks like there's more going on in that build failure:

* There's a circular require due to `smells.rb` requiring all files in `smells/`.
* One sample cuke is failing due to different smells being detected. That one seems odd.

**Comment by troessner** [comment]
>> There's a circular require due to smells.rb requiring all files in smells/.

I wonder why this works locally but not on travis. I cant see anything specific in what travis does and in the end it just calls "bundle exec rake" which works for me across multiple ruby versions.

Regarding the problem:

I'd suggest moving all smell types into smells/smell_types, require those in smell_repository and get rid of smells.rb. This also feels like a clean architecture to me.

What do you think about that?

**Comment by mvz** [comment]
> I'd suggest moving all smell types into smells/smell_types, require those in smell_repository and get rid of smells.rb. This also feels like a clean architecture to me.

Perhaps we can start by moving `smell-repository.rb` up to root level? It seems the simpler solution for now. Otherwise, I think `smells/` will end up containing just `smell_repository.rb` and `smell_types/`?

The Travis failures are caused by newer versions of `parser` by the way. I don't know yet what the relevant change is, but it all works fine with `2.2.2.2` but breaks with `2.2.2.5`. :frowning:

**Comment by mvz** [comment]
The problem in parser `2.2.2.5` is probably due to https://github.com/whitequark/parser/issues/194.

**Comment by troessner** [comment]
Then let's pin parser to 2.2.2.2 for now until this issue is fixed. 
Objections?
And afterwards we should really discuss pinning all the gems we use to the version that is currently running. This is not the first time we're having this issue.

**Comment by mvz** [comment]
:+1: on pinning parser.

**Comment by chastell** [comment]
> Then let's pin parser to 2.2.2.2 for now until this issue is fixed.

This makes sense. Note: from a cursory look at the issue it seems like a _feature_ of parser (different comment association), so we’ll hav to _fix_ it on our side…

> And afterwards we should really discuss pinning all the gems we use to the version that is currently running. This is not the first time we're having this issue.

Hm, I think semver exists for a reason and most dependencies actually follow it – I don’t think parser does, though. We should definitely pin parser.

**Comment by mvz** [comment]
@troessner can you rebase to see if the tests pass now?

**Comment by troessner** [comment]
They're passing!
